Mathlib/AlgebraicGeometry/Morphisms/Separated.lean

```diff
@@ -113,6 +114,83 @@ instance [IsSeparated g] :
+  let i := 𝒰.f (f.base ((pullback.fst f f).base x))
+  obtain ⟨y : 𝒰.obj i, hy : (𝒰.map i).base y = f.base _⟩ :=
+    𝒰.covers (f.base ((pullback.fst f f).base x))
+  obtain ⟨z, hz₁, hz₂⟩ := exists_preimage_pullback _ _ hy.symm
+  let j := (𝒱 i).f z
+  obtain ⟨w : (𝒱 i).obj j, hy : ((𝒱 i).map j).base w = z⟩ := (𝒱 i).covers z
+  refine ⟨i, j, ?_⟩
+  simp_rw [diagonalCover_map]
+  show x ∈ Set.range _
+  erw [range_map]
+  simp only [comp_coeBase, TopCat.coe_comp, Set.mem_inter_iff, Set.mem_preimage, Set.mem_range,
+    Function.comp_apply, ← H, and_self, ← hz₁, ← hy]
+  exact ⟨w, rfl⟩
+
+lemma Scheme.Pullback.range_diagonal_subset_diagonalCoverDiagonal :
+    Set.range (pullback.diagonal f).base ⊆ diagonalCoverDiagonal f 𝒰 𝒱 := by
+  rintro _ ⟨x, rfl⟩
+  simp only [diagonalCoverDiagonal, openCoverOfBase_J, openCoverOfBase_obj, openCoverOfLeftRight_J,
+    Opens.iSup_mk, Opens.carrier_eq_coe, Hom.opensRange_coe, Opens.coe_mk, Set.mem_iUnion,
+    Set.mem_range, Sigma.exists]
+  let i := 𝒰.f (f.base x)
+  obtain ⟨y : 𝒰.obj i, hy : (𝒰.map i).base y = f.base x⟩ := 𝒰.covers (f.base x)
+  obtain ⟨z, hz₁, hz₂⟩ := exists_preimage_pullback _ _ hy.symm
+  let j := (𝒱 i).f z
+  obtain ⟨w : (𝒱 i).obj j, hy : ((𝒱 i).map j).base w = z⟩ := (𝒱 i).covers z
+  refine ⟨i, j, (pullback.diagonal ((𝒱 i).map j ≫ pullback.snd f (𝒰.map i))).base w, ?_⟩
+  rw [← hz₁, ← hy, ← Scheme.comp_base_apply, ← Scheme.comp_base_apply]
+  erw [← Scheme.comp_base_apply]
+  congr 4
+  apply pullback.hom_ext <;> simp [diagonalCover_map, OpenCover.pullbackHom]
+
+lemma isClosedImmersion_diagonal_restrict_diagonalCoverDiagonal
+    [∀ i, IsAffine (𝒰.obj i)] [∀ i j, IsAffine ((𝒱 i).obj j)] :
+    IsClosedImmersion (pullback.diagonal f ∣_ diagonalCoverDiagonal f 𝒰 𝒱) := by
+  let U : (Σ i, (𝒱 i).J) → (diagonalCoverDiagonal f 𝒰 𝒱).toScheme.Opens := fun i ↦
+    (diagonalCoverDiagonal f 𝒰 𝒱).ι ⁻¹ᵁ ((diagonalCover f 𝒰 𝒱).map ⟨i.1, i.2, i.2⟩).opensRange
+  have hU (i) : (diagonalCoverDiagonal f 𝒰 𝒱).ι ''ᵁ U i =
+      ((diagonalCover f 𝒰 𝒱).map ⟨i.1, i.2, i.2⟩).opensRange := by
+    rw [TopologicalSpace.Opens.functor_obj_map_obj, inf_eq_right, Hom.image_top_eq_opensRange,
+      Opens.opensRange_ι]
+    exact le_iSup (fun i : Σ i, (𝒱 i).J ↦ ((diagonalCover f 𝒰 𝒱).map ⟨i.1, i.2, i.2⟩).opensRange) i
+  have hf : iSup U = ⊤ :=
+    (TopologicalSpace.Opens.map_iSup _ _).symm.trans (diagonalCoverDiagonal f 𝒰 𝒱).ι_preimage_self
+  rw [IsLocalAtTarget.iff_of_iSup_eq_top (P := @IsClosedImmersion) _ hf]
+  intro i
+  rw [MorphismProperty.arrow_mk_iso_iff (P := @IsClosedImmersion) (morphismRestrictRestrict _ _ _),
+    MorphismProperty.arrow_mk_iso_iff (P := @IsClosedImmersion) (morphismRestrictEq _ (hU i)),
+    MorphismProperty.arrow_mk_iso_iff (P := @IsClosedImmersion) (diagonalRestrictIsoDiagonal ..)]
+  infer_instance
+
+@[stacks 0DVA]
+lemma isSeparated_of_injective (hf : Function.Injective f.base) :
+    IsSeparated f := by
```

Comment:
Glad to see [this](https://github.com/stacks/stacks-project/commit/4b2e2ea9e54c787fafd85df59a0d13a367b3d5d2) formalized!